app/cdap/components/PipelineDiff/DiffCanvas/index.tsx

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import React, { ComponentType } from 'react';
+import {
+  Node,
+  Background,
+  Controls,
+  Edge,
+  MiniMap,
+  ReactFlow,
+  ReactFlowProvider,
+  NodeProps,
+} from 'reactflow';
+import { DefaultPluginNode, PluginNodeWithAlertError } from './PluginNode';
+import { getReactflowPipelineGraph } from '../util/reactflowGraph';
+import { AvailablePluginsMap, IPipelineConfig, IPipelineDiffMap } from '../types';
+import { IPipelineNodeData } from '../store/diffSlice';
+import { pluginReactflowNodeType } from '../util/helpers';
+
+type NodeType = ReturnType<typeof pluginReactflowNodeType>;
+const nodeTypes: Record<NodeType, ComponentType<NodeProps>> = {
```

Comment:
Is this reactflow's recommended ways of defining nodeTypes?

Reply:
I did not see a recommended type usage for it, however there was an example where they used the `NodeTypes` type. The difference with the way it is currently typed (see updated code) is that it is more specific on which node types exist within the application. I think this is more type safe, but I am open to changing it if it is an unnecessary complication.